root/usr/lib64/ruby/vendor_ruby/discovery/menu.rb

```diff
@@ -191,15 +209,22 @@ def main_loop
       active_screen = :screen_welcome
     end
   end
+
+  # the pipeline loop
   while ! [:quit].include? active_screen
-    if active_screen.is_a?(Array)
-      log_debug "Entering #{active_screen[0]}"
-      active_screen = send(*active_screen)
-    else
-      log_debug "Entering #{active_screen}"
-      active_screen = send(active_screen)
-    end
+    log_debug "Entering #{active_screen}"
+    next_screen = send(active_screen, pipeline)
     Newt::Screen.pop_window()
+    active_screen = pipeline.next
+    unless active_screen
+      log_debug "No screens in pipeline, fallback to welcome"
+      active_screen = :screen_welcome
+    end
+    if active_screen.is_a? Array
+      result = next_screen.to_i
+      log_debug "Result is #{result}, picking out of #{active_screen.inspect}"
+      active_screen = active_screen[result]
+    end
```

Comment:
The main loop - it takes "next screen" variable from the pipeline and enters it. Over and over again until "exit".